### ISSUE TITLE
Is it possible to provide state+action in the policy input and get Q(state, action) in the output?

### ISSUE THREAD

**Issue by cgr71ii** [DLR-RM/stable-baselines3#1809]
### ❓ Question

Hi!

I'm trying to solve a problem related to focused crawling (similar to https://inria.hal.science/hal-01851547/file/crawling_2018.pdf). The problem is that for each downloaded document the action space changes: it is unknown beforehand and unbounded. Then I thought about applying something similar to a per-action value of the action value (as applied in figure 1b in https://aclanthology.org/P16-1153.pdf). Is it possible to do this with stable baselines3? I know the standard approach is to have |state_space| input and |action_space| output, but can I apply the policy as many times as there are actions for a given state and compute the action value (i.e. provide state+action in input and get Q(state, action) in output) in e.g. DQN?

Thank you!

### Checklist

- [X] I have checked that there is no similar [issue](https://github.com/DLR-RM/stable-baselines3/issues) in the repo
- [X] I have read the [documentation](https://stable-baselines3.readthedocs.io/en/master/)
- [X] If code there is, it is [minimal and working](https://github.com/DLR-RM/stable-baselines3/issues/982#issuecomment-1197044014)
- [X] If code there is, it is formatted using the [markdown code blocks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks) for both code and stack traces.

**Comment by araffin** [comment]
Duplicate of https://github.com/DLR-RM/stable-baselines3/issues/568 and https://github.com/DLR-RM/stable-baselines3/issues/1257 and https://github.com/DLR-RM/stable-baselines3/issues/1347

**Comment by cgr71ii** [comment]
Hi!

Well, I took a look at the algorithms you mentioned in this repo, and it's not what I meant. These algorithms might fit my situation if I were looking for a similar embedding of my coded action, but what I want is the exact action to execute according to the algorithm. In the code I observed the relevant points are:

1. https://github.com/DLR-RM/stable-baselines3/blob/a9273f968eaf8c6e04302a07d803eebfca6e7e86/stable_baselines3/td3/policies.py#L57 - It calculates the number of output values for the ANN
2. https://github.com/DLR-RM/stable-baselines3/blob/a9273f968eaf8c6e04302a07d803eebfca6e7e86/stable_baselines3/common/preprocessing.py#L197 - For the continuous actions, the multiplication of the shape is returned
3. https://github.com/DLR-RM/stable-baselines3/blob/a9273f968eaf8c6e04302a07d803eebfca6e7e86/stable_baselines3/td3/policies.py#L58 - The input of the ANN is created according to only the state and not the state+action

I can use these algorithms if I represent my actions using continuous values, but then I'd need to use something like kNN to select the specific action. I'll give it a try anyway.

Thanks for your help!

**Comment by cgr71ii** [comment]
To anyone with the same problem, this paper might help: https://arxiv.org/abs/1512.07679. Also I implemented the Wolpertinger policy and the few changes in the DDPG training algorithm mentioned in the paper: https://github.com/cgr71ii/stable-baselines3/tree/wolpertinger

I found different repos with the Wolpertinger policy implementation, but it seems that they did not include the changes in the DDPG algorithm (e.g. https://github.com/jimkon/Deep-Reinforcement-Learning-in-Large-Discrete-Action-Spaces/issues/36).
